src/udev/udev-builtin-net_id.c

```diff
@@ -810,6 +810,8 @@ static int names_devicetree(sd_device *dev, const char *prefix, EventMode mode)
          * devicetree_syspath = /sys/firmware/devicetree/base
          * ofnode_syspath = /sys/firmware/devicetree/base/soc/ethernet@deadbeef
          * ofnode_path = soc/ethernet@deadbeef
+         * eth_ports_child_syspath = /sys/firmware/devicetree/base/bus@10000/ethernet@deadbeef/ethernet-ports/port@1
+         * eth_ports_child_path = /bus@10000/ethernet@deadbeef/ethernet-ports/port@1
```

Comment:
I don't really understand the paths here. In the case of switch ports, shouldn't the paths be something like

```
dev = /sys/bus/platform/devices/deadbeef.ethernet/1.port/netdev
devicetree_syspath = /sys/firmware/devicetree/base
ofnode_syspath = /sys/firmware/devicetree/base/soc/ethernet@deadbeef/ethernet-ports/port@1
ofnode_path = soc/ethernet@deadbeef/ethernet-ports/port@1
```

? Otherwise how do we know which port we have?

Reply:
Sorry I don't think I fully understood the question, but please take as example the verdin-am62 devicetree: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm64/boot/dts/ti/k3-am62-verdin.dtsi#n24

```c
ethernet0 = &cpsw_port1;
ethernet1 = &cpsw_port2;
```

The aliases will point to these ethernet switch ports (https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm64/boot/dts/ti/k3-am62-main.dtsi#n730):

```c
            ethernet-ports {
			#address-cells = <1>;
			#size-cells = <0>;

			cpsw_port1: port@1 {
				reg = <1>;
				ti,mac-only;
				label = "port1";
				phys = <&phy_gmii_sel 1>;
				mac-address = [00 00 00 00 00 00];
				ti,syscon-efuse = <&wkup_conf 0x200>;
			};

			cpsw_port2: port@2 {
				reg = <2>;
				ti,mac-only;
				label = "port2";
				phys = <&phy_gmii_sel 2>;
				mac-address = [00 00 00 00 00 00];
			};
		};
```

Therefore the code will loop trought the child nodes of `ethernet-ports` (or `ports`) and compare with the ethernet aliases path.

On the AM62 SoC the ethernet switch is connect to the bus called `cbass_main` on the devicetree here: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm64/boot/dts/ti/k3-am65.dtsi#n50

Reply:
Yes, I understand the devicetree side. However, in sysfs the `netdev` should be a direct child of the port. So why do we need to do anything special to look up the alias? The existing code should find the aliases normally.

Reply:
Ok understood. With the current code, I added the following print debugging:

```c
                if (sd_device_get_sysattr_value(aliases_dev, alias, &alias_path) < 0)
                        continue;
                log_device_debug(dev, "hfranco: alias_path: %s", alias_path);
                log_device_debug(dev, "hfranco: ofnode_path: %s", ofnode_path);

                if (!path_equal(ofnode_path, alias_path))
                        continue;
                log_device_debug(dev, "hfranco: paths are equal");
```

And with this devicetree, this is the result:

```bash
(udev-worker)[235]: eth1: hfranco: alias_path: /bus@f0000/ethernet@8000000/ethernet-ports/port@2
(udev-worker)[235]: eth1: hfranco: ofnode_path: /bus@f0000/ethernet@8000000
```

Which fails at `if (!path_equal(ofnode_path, alias_path))`.  These are the syspath and devicetree_syspath:

```bash
(udev-worker)[235]: eth1: hfranco: syspath: /sys/firmware/devicetree/base/bus@f0000/ethernet@8000000
(udev-worker)[235]: eth1: hfranco: devicetree_syspath: /sys/firmware/devicetree/base
```

So later I just added the `FOREACH_DEVICE_CHILD(eth_ports_dev, child) {` to match against the aliases.

Please let me know if something is not correct.

Reply:
OK, so the issue is that there can be multiple netdevs with the same parent device. Your approach will not work because we need a way to disambiguate between all the different netdevs a device may have registered. Without that, all the netdevs will end up with the same alias (for e.g. port@0).

Can we use the `dev_id` or `dev_port` files to determine which port we have? Failing that, what about devlink?

Reply:
(I promise I'm done editing the above comment)

Reply:
Got it @sean-anderson-seco, thanks for the comment.

I was trying to debug and update the code but I dont think there is a `of_node` of `net/<interface>/of_node`, am I correct? This is what I see on my device:

```bash
root@verdin-am62-15133532:~# find /sys/ -name *of_node* | grep ethernet
/sys/devices/platform/bus@f0000/8000000.ethernet/of_node
/sys/devices/platform/bus@f0000/8000000.ethernet/8000f00.mdio/of_node
/sys/devices/platform/bus@f0000/8000000.ethernet/8000f00.mdio/mdio_bus/8000f00.mdio/of_node
/sys/devices/platform/bus@f0000/8000000.ethernet/8000f00.mdio/mdio_bus/8000f00.mdio/8000f00.mdio:00/of_node
/sys/devices/platform/bus@f0000/8000000.ethernet/8000f00.mdio/mdio_bus/8000f00.mdio/8000f00.mdio:07/of_node
```

```bash
root@verdin-am62-15133532:~# ls /sys/devices/platform/bus@f0000/8000000.ethernet/net/
enx00142de6eb5c  enx00142df6eb5c
root@verdin-am62-15133532:~# ls /sys/devices/platform/bus@f0000/8000000.ethernet/net/enx00142de6eb5c/
addr_assign_type      carrier_up_count      gro_flush_timeout     napi_defer_hard_irqs  power                 threaded
addr_len              dev_id                ifalias               netdev_group          proto_down            tx_queue_len
address               dev_port              ifindex               operstate             queues                type
broadcast             device                iflink                phydev                speed                 uevent
carrier               dormant               link_mode             phys_port_id          statistics
carrier_changes       duplex                mtu                   phys_port_name        subsystem
carrier_down_count    flags                 name_assign_type      phys_switch_id        testing
```

In this case my devs are getting the mac address naming assign.
Can we use a different file instead of `of_node`?

Reply:
Hm, interesting. I have a device with e.g.

```
/sys/module/sja1105/drivers/spi:sja1105/spi2.0/net/swp0/of_node
/sys/module/sja1105/drivers/spi:sja1105/spi2.0/net/swp1/of_node
/sys/module/sja1105/drivers/spi:sja1105/spi2.0/net/swp2/of_node
/sys/module/sja1105/drivers/spi:sja1105/spi2.0/net/swp3/of_node
```

Well, I suppose the alternative is to use `phys_port_name`. But we need to parse it (c.f. `__devlink_port_phys_port_name_get`).

Reply:
@sean-anderson-seco 

> But given that example, maybe this is just the wrong approach? Do we really need to use aliases for the ports

In the end these are on-board, or better in-SoC, ethernet devices, and they would fit in the definition you had in the patch that introduced support for DT ethernet aliases.

To me, it feels that honoring the ethernet alias is still a good idea here, for the same reasons you wrote in the past.

Reply:
Yes, but without an `of_node` it is difficult to determine which alias to use. I think you have the following approaches:

- Update Linux to create an `of_node` symlink like is done for DSA switches and patch `systemd` to look for this `of_node` in the netdev directory.
- Create a new naming scheme using the `phys_switch_id` and `phy_port_name` files.

Reply:
All right, thanks @sean-anderson-seco for the comments, I will take a look at the kernel code and check if I can propose a patch, then we can come back to this PR.

Reply:
Hi @sean-anderson-seco, can you please share where the of_node links are pointing to on your device? Maybe a `ls -lh /sys/modules/sja1105/drivers/spi:sja1105/spi2.0/net/swp*/of_node`

```
/sys/module/sja1105/drivers/spi:sja1105/spi2.0/net/swp0/of_node
/sys/module/sja1105/drivers/spi:sja1105/spi2.0/net/swp1/of_node
/sys/module/sja1105/drivers/spi:sja1105/spi2.0/net/swp2/of_node
/sys/module/sja1105/drivers/spi:sja1105/spi2.0/net/swp3/of_node
```

The patch I proposed to the kernel does not create the of_node's inside the `net` folder, only into the platform, so I am a bit confused about where these of_node's are coming from.

Reply:
Adding more information to this issue, I switched my hardware to a NXP i.MX8MP and this one does not have any ethernet switch, instead, it has two different ethernet nodes in different addresses.

In this case, the device tree naming scheme from systemd is correct (as expected):

```
root@verdin-imx8mp-15128376:~# ip a | grep end
2: end1: <NO-CARRIER,BROADCAST,MULTICAST,DYNAMIC,UP> mtu 1500 qdisc mq state DOWN group default qlen 1000
5: end0: <BROADCAST,MULTICAST,DYNAMIC,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    inet 10.22.1.184/24 brd 10.22.1.255 scope global end0
```

But this one does not create any "of_node" files inside the net folder too:

```
root@verdin-imx8mp-15128376:~# find /sys/devices/platform/soc\@0/30800000.bus/30be0000.ethernet/net/ -name of_node
root@verdin-imx8mp-15128376:~# find /sys/devices/platform/soc\@0/30800000.bus/30bf0000.ethernet/net/ -name of_node
root@verdin-imx8mp-15128376:~# ls /sys/devices/platform/soc\@0/30800000.bus/*/net/
/sys/devices/platform/soc@0/30800000.bus/30be0000.ethernet/net/:
end1

/sys/devices/platform/soc@0/30800000.bus/30bf0000.ethernet/net/:
end0
```

By checking the sja1105 driver I could not find where this of_node is being created and why it is not being created for other drivers, I am afraid maybe this sja1105 is somehow different than the others, and this approach to use the of_node might not work for the ethernet-switch... not sure.

Reply:
Hi, sorry for the long delay.

I believe the way this is set is first `dsa_port_parse_of` sets `dp->dn`:

```
dsa_register_switch
dsa_switch_probe
dsa_switch_parse_of
dsa_switch_parse_ports_of
dsa_port_parse_of /* dp->dn = dn; */
```

Then, the `of_node` link is created by `device_add_class_symlinks` via `register_netdevice`.

```
dsa_register_switch
dsa_switch_probe
dsa_tree_setup
dsa_tree_setup_ports
dsa_port_setup
dsa_slave_create /* slave_dev->dev.of_node = port->dn; */
register_netdevice
netdev_register_kobject
device_add
device_add_class_symlinks
```

So I believe you can add this to your device by saving `port_np` in `am65_cpsw_nuss_init_slave_ports` and setting  setting `ndev->dev._of_node` to it in `am65_cpsw_nuss_init_port_ndev`.